ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -24,28 +25,116 @@
+  // the version to indicate the serialized format.
+  private Short version;
+  private Byte flag;
+  private Byte datacenterId;
+  private Short accountId;
+  private Short containerId;
   private PartitionId partitionId;
   private String uuid;
-  private static int UUID_Size_In_Bytes = 4;
 
   /**
-   * Constructs a new unique BlobId for the specified partition.
+   * Constructs a new unique BlobId. This will construct a blob in version 2, so the serialization embeds accountId,
+   * containerId, datacenterId, and partitionId.
+   * @param flag A byte to embed additional information of this blobId.
+   * @param datacenterId The id of the datacenter to be embedded into the blob.
+   * @param accountId The id of the {@link Account} to be embedded into the blob.
+   * @param containerId The id of the {@link Container} to be embedded into the blob.
+   * @param partitionId The partition where this blob is to be stored.
+   */
+  BlobId(Byte flag, Byte datacenterId, Short accountId, Short containerId, PartitionId partitionId) {
+    this.flag = flag;
+    this.datacenterId = datacenterId;
+    this.accountId = accountId;
+    this.containerId = containerId;
+    this.partitionId = partitionId;
+    uuid = UUID.randomUUID().toString();
+    populateMissingFieldsAndDetermineVersion();
+  }
+
+  /**
+   * Re-constructs existing blobId by deserializing from data input stream. This constructor includes an optional check
+   * that the stream has no more available bytes after reading.
    *
-   * @param partitionId of Partition in which blob is to be stored.
+   * @param stream from which to deserialize the blobid
+   * @param clusterMap of the cluster that the partition of the blob id belongs to
+   * @param ensureFullyRead {@code true} if the stream should have no more available bytes after deserializing the blob
+   *                        ID.
+   * @throws IOException
    */
-  public BlobId(PartitionId partitionId) {
+  private BlobId(DataInputStream stream, ClusterMap clusterMap, boolean ensureFullyRead) throws IOException {
+    this.version = stream.readShort();
+    switch (version) {
+      case BLOB_ID_V1:
+        flag = DEFAULT_FLAG;
+        datacenterId = LEGACY_DATACENTER_ID;
+        accountId = LEGACY_ACCOUNT_ID;
+        containerId = LEGACY_CONTAINER_ID;
+        break;
+      case BLOB_ID_V2:
+        flag = stream.readByte();
+        datacenterId = stream.readByte();
+        accountId = stream.readShort();
+        containerId = stream.readShort();
+        break;
+      default:
+        throw new IllegalArgumentException("blob id version " + version + " not supported for blob id");
+    }
+    partitionId = clusterMap.getPartitionIdFromStream(stream);
     if (partitionId == null) {
       throw new IllegalArgumentException("Partition ID cannot be null");
     }
-    this.partitionId = partitionId;
-    this.uuid = UUID.randomUUID().toString();
+    uuid = Utils.readIntString(stream);
```

Comment:
I see that you have changed the UUID_SIZE to short. Then this has to be short followed by a string right.